db_store/src/iam_auth_pool.rs

```diff
@@ -0,0 +1,227 @@
+use crate::{error::invalid_configuration, Error, Result, Settings};
+use sqlx::{
+    postgres::{PgConnectOptions, Postgres},
+    Pool,
+};
+
+use aws_credential_types::Credentials;
+use aws_sig_auth::signer::{self, HttpSignatureType, OperationSigningConfig, RequestConfig};
+use aws_smithy_http::body::SdkBody;
+use aws_types::region::{Region, SigningRegion};
+use aws_types::SigningService;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+pub async fn connect(
+    settings: &Settings,
+    shutdown: triggered::Listener,
+) -> Result<(Pool<Postgres>, futures::future::BoxFuture<'static, Result>)> {
+    let aws_config = aws_config::load_from_env().await;
+    let client = aws_sdk_sts::Client::new(&aws_config);
+    let connect_parameters = ConnectParameters::try_from(settings)?;
+    let connect_options = connect_parameters.connect_options(&client).await?;
+
+    let pool = settings
+        .pool_options()
+        .connect_with(connect_options)
+        .await?;
+
+    let cloned_pool = pool.clone();
+    let join_handle =
+        tokio::spawn(async move { run(client, connect_parameters, cloned_pool, shutdown).await });
+
+    Ok((
+        pool,
+        Box::pin(async move {
+            match join_handle.await {
+                Ok(Ok(())) => Ok(()),
+                Ok(Err(err)) => Err(err),
+                Err(err) => Err(Error::from(err)),
+            }
+        }),
+    ))
+}
+
+async fn run(
+    client: aws_sdk_sts::Client,
+    connect_parameters: ConnectParameters,
+    pool: Pool<Postgres>,
+    shutdown: triggered::Listener,
+) -> Result {
+    let duration = std::time::Duration::from_secs(connect_parameters.iam_duration_seconds as u64)
+        - Duration::from_secs(120);
+
+    loop {
+        let shutdown = shutdown.clone();
+
+        tokio::select! {
+            _ = shutdown => break,
+            _ = tokio::time::sleep(duration) => {
+                let connect_options = connect_parameters.connect_options(&client).await?;
+                pool.set_connect_options(connect_options);
```

Comment:
Not sure I understand the question, but if the pool wants to reconnect it will use the `ConnectOptions` to do so.  This overwrites the `ConnectOptions` the pool has before the auth token expires, so it should always have a valid auth token and can reconnect whenever it wants.

Reply:
oh, so the pool will always try to reconnect and this just ensures that it can?